### ISSUE TITLE
Bug in `git submodule update --recursive --init`?

### ISSUE THREAD

**Issue by akawashiro** [seeraven/gitcache#5]
Thank you for publishing great software.

I tried to use `gitcache` for my work and I found `git submodule update --recursive --init` doesn't work. Here is a reproducing script and its result. Do you have any ideas on this behavior?

```
> cat gitcache-val.sh
#! /bin/bash -eu

REAL_GIT_DIR1=real_git_dir1
REAL_GIT_DIR2=real_git_dir2
GITCACHE_DIR1=gitcache_dir1
GITCACHE_DIR2=gitcache_dir2

rm -rf ${REAL_GIT_DIR1}
rm -rf ${REAL_GIT_DIR2}
rm -rf ${GITCACHE_DIR1}
rm -rf ${GITCACHE_DIR2}

mkdir ${REAL_GIT_DIR1}
pushd ${REAL_GIT_DIR1}
/usr/bin/git clone https://github.com/aws/aws-sdk-cpp.git --single-branch --branch 1.9.188
pushd aws-sdk-cpp
/usr/bin/git submodule update --recursive --init
popd
popd

mkdir ${REAL_GIT_DIR2}
pushd ${REAL_GIT_DIR2}
/usr/bin/git clone https://github.com/aws/aws-sdk-cpp.git --single-branch --branch 1.9.188
pushd aws-sdk-cpp
/usr/bin/git submodule update --recursive --init
popd
popd

mkdir ${GITCACHE_DIR1}
pushd ${GITCACHE_DIR1}
git clone https://github.com/aws/aws-sdk-cpp.git --single-branch --branch 1.9.188
pushd aws-sdk-cpp
git submodule update --recursive --init
popd
popd

mkdir ${GITCACHE_DIR2}
pushd ${GITCACHE_DIR2}
git clone https://github.com/aws/aws-sdk-cpp.git --single-branch --branch 1.9.188
pushd aws-sdk-cpp
git submodule update --recursive --init
popd
popd

echo ========== Number of diff of real_git_dir1 and real_git_dir2 ==========
diff -rq real_git_dir1/aws-sdk-cpp real_git_dir2/aws-sdk-cpp | grep -v "[.]git" | wc -l
echo ========== Number of diff of real_git_dir1 and gitcache_dir1 ==========
diff -rq real_git_dir1/aws-sdk-cpp gitcache_dir1/aws-sdk-cpp | grep -v "[.]git" | wc -l
echo ========== Number of diff of gitcache_dir1 and gitcache_dir2 ==========
diff -rq gitcache_dir1/aws-sdk-cpp gitcache_dir2/aws-sdk-cpp | grep -v "[.]git" | wc -l

> ./gitcache-val.sh
========== Number of diff of real_git_dir1 and real_git_dir2 ==========
0
========== Number of diff of real_git_dir1 and gitcache_dir1 ==========
218
========== Number of diff of gitcache_dir1 and gitcache_dir2 ==========
0
```

**Comment by akawashiro** [comment]
I think `git submodule update` has some bug. This small patch fix the result of `gitcache-val.sh`.

```
> git diff --patch
diff --git a/git_cache/git_command.py b/git_cache/git_command.py
index 45eca11..a83d804 100644
--- a/git_cache/git_command.py
+++ b/git_cache/git_command.py
@@ -109,8 +109,9 @@ def handle_git_command(called_as, args):
     elif git_options.get_command() == 'submodule_init':
         sys.exit(git_submodule_init(git_options))
 
-    elif git_options.get_command() == 'submodule_update':
-        sys.exit(git_submodule_update(called_as, git_options))
+    # TODO(akawashiro):
+    # elif git_options.get_command() == 'submodule_update':
+    #     sys.exit(git_submodule_update(called_as, git_options))
 
     LOG.debug("Command '%s' is not handled by gitcache. Calling the real git command.",
               git_options.get_command())
```

**Comment by seeraven** [comment]
Hi!
Thank you for the bug report. I suspect the problem might be caused by the `--recursive` option, but I am not really sure about it. I have to admit, that I've never used the submodules feature of git in a recursive way.
The reason it works by commenting out the `submodule update` wrapping is simple, as it then uses the real git command twice. ;-)
I will take your example and start with adding a new functional test, so I can test the bugfix against it.

**Comment by akawashiro** [comment]
Thank you for your quick response.

**Comment by seeraven** [comment]
Hello again!

I implemented the recursive checkout and tested it with your example. So far it works, the only difference is now in a single repository that was checked out on master but was then reverted to an older commit, leaving a submodule checkout. This is then reported correctly as a difference. I don't think it hurts with this particular example, but it shows the deficit with the current submodule handling within gitcache: gitcache clones or fetches the submodules itself, leaving the .git directories in the individual directories. The real submodule command moves the .git directories into the one of the parent repository and have access to valuable information like the commit hash to be checked out of the individual submodules. Unfortunately, I have not found an easy way of hooking into the submodule commands for clone/update.

The version with the bugfix can be found on https://github.com/seeraven/gitcache/releases/tag/v1.0.7. I would be very happy if your can test it and tell me whether this works for you.
